Add support for Rust to the CI builds.

d3zd3z opened the pull request:
The rust toolchain is already in the path, it just needs to be added to the path.

There is also a workaround to install the target support, which is needed until these are added to the docker images.

d3zd3z committed:
rust: Add initial target list

Add the target list early, as this is needed by the workflow in order to
know what targets to add rust support for.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
ci: workflow: Add rust targets to twister

Add the Rust tools to the path in the twister build.  Also add any rust
targets that are needed.  The docker CI image should have these already
installed, but this will close the window between when new targest are
added and when the docker image can be updated.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
I'm not quite sure of the proper procedure here. I have tested what this script does locally, but it appears from the build results that CI in this case is going to run the workflows from `main` and not the changes in this PR. This will need to be merged into main before #75904 will be able to pass CI.

fabiobaltieri commented on .github/workflows/twister.yaml:
```diff
@@ -188,12 +188,14 @@ jobs:
           west config --global update.narrow true
           west update --path-cache /repo-cache/zephyrproject 2>&1 1> west.update.log || west update --path-cache /repo-cache/zephyrproject 2>&1 1> west.update.log || ( rm -rf ../modules ../bootloader ../tools && west update --path-cache /repo-cache/zephyrproject)
           west forall -c 'git reset --hard HEAD'
+          for target in `cat lib/rust/targets.txt`; do rustup target install $target; done
```
do you expect having more rustup commands in the future? wondering if it would make sense to have these two in their own step, like having a "Rust Environment Setup", just to keep them isolated, I'd imagine this blocked would have to be copypasted to other workflows as well

fabiobaltieri replied:
Also, isn't there a standard actions for doing this? I see few on github though they all look community maintained.

nashif replied:
> do you expect having more rustup commands in the future? wondering if it would make sense to have these two in their own step

agree, lets keep those in own step, later to be removed when the docker has all the toolchains.

d3zd3z replied:
Yes, there will likely be another step, to pre-fetch dependencies before starting the build.

But, moving to a different step makes me leery of what I feel like I can write without any way to test it. Any idea how to actually test this without merging it?

fabiobaltieri replied:
Oh because `pull_request_target` runs on the previous version? I would push it on https://github.com/zephyrproject-rtos/zephyr-testing/

nashif replied:
now that I think about it, I will be against such a change which pulls toolchains in each action invocation. I would instead focus on doing it in the docker and get this done.
Pull packages and toolchains each time we run the twister workflow both will be slower and error-prone, i.e. we will be pulling data from the network, something that can fail...

so https://github.com/zephyrproject-rtos/docker-image/pull/194 is the right approach, we just need to get CI working on that one, not sure what is going on. @stephanosio ??

fabiobaltieri replied:
> now that I think about it, I will be against such a change which pulls toolchains in each action invocation

You could cache it though, it's no different than the python packages.

teburd replied:
rustup is quite reliable, python packages have broken for me in the past when python versions change sadly

d3zd3z replied:
I'm ok with having the toolchains in the docker image. Keep in mind that if the `rustup` here pulls in the same set that are in the docker image, the command will do nothing, it just prints a message that it is up to date.

I was kind of intending this to be able to break the dependency between the docker image and the Zephyr tree. If in the future, we add another target that is supported, it would automatically be installed until docker could be updated. Otherwise, adding the target will depend on getting the docker image update through.

d3zd3z replied:
Who can I ask about push permission on the zephyr-testing repo?

fabiobaltieri replied:
> Who can I ask about push permission on the zephyr-testing repo?

@stephanosio @nashif? Not sure how it works, I think I have it from the release team one but maybe they can set a branch up, or maybe you can get a collab- here just for that

d3zd3z committed:
ci: workflow: Separate rust workflow setup

The workflow still needs the environment in a separate step so that the
path will have cargo in it.

Signed-off-by: David Brown <david.brown@linaro.org>

fabiobaltieri commented on .github/workflows/twister.yaml:
```diff
@@ -188,6 +188,7 @@ jobs:
             git log  --pretty=oneline | head -n 10
           fi
           echo "$HOME/.local/bin" >> $GITHUB_PATH
+          echo "$HOME/.cargo/bin" >> $GITHUB_PATH
```
I'd move this as well, keep the steps isolated at some extent, you don't need it here, right?

d3zd3z replied:
It needs to be in a separate run step from the rustup command below. This won't get removed, even if the rustup target add command does.

d3zd3z commented:
I have pushed this to testing, and the twister run all passes, so if we're ok to merge this, it is ready. Are we ok with it going through this way, or should I make the environment line a separate run block?

d3zd3z commented:
https://github.com/zephyrproject-rtos/zephyr-testing/actions/runs/10356594071

cfriedt reviewed (approved):
LGTM

nashif reviewed (changes requested):
I am not a fan of pulling toolchains in our most frequently run workflow without actually using them. Prefer this to be done either using the docker image or with actual users, i.e. the rust feature in place, to make this happen, I suggest to get everything integrated and working in a collab branch dedicated for rust.

d3zd3z commented:
Closing as we are going through the rust-collab branch.
